File: tutorial/js/02_read_insert_and_delete.mjs
/**
 * Example 2: Read rows, insert a new row, delete a row by _rowid.
 *
 * Default target: https://dbserver.streamlit.app
 * Prerequisites: run 01_create_table_and_write.mjs first
 *
 * Run: node tutorial/js/02_read_insert_and_delete.mjs
 */
import {
  ACTOR,
  DATABASE,
  apiFetch,
  apiUrl,
  checkResponse,
  siteHeaders,
  writeHeaders,
} from "./config.mjs";

const TABLE = "tutorial_products";

async function listRows(limit = 20) {
  const url = new URL(apiUrl(`/api/databases/${DATABASE}/tables/${TABLE}/rows`));
  url.searchParams.set("limit", String(limit));
  url.searchParams.set("order", "asc");
  const resp = await apiFetch(url, { headers: siteHeaders() });
  return checkResponse(resp, "list rows");
}

async function insertRow(row) {
  const resp = await apiFetch(apiUrl(`/api/databases/${DATABASE}/tables/${TABLE}/rows`), {
    method: "POST",
    headers: writeHeaders(),
    body: JSON.stringify({ actor: ACTOR, ...row }),
  });
  return checkResponse(resp, "insert row");
}

async function deleteRow(rowId) {
  const resp = await apiFetch(
    apiUrl(`/api/databases/${DATABASE}/tables/${TABLE}/rows/${rowId}`),
    {
      method: "DELETE",
      headers: writeHeaders(),
      body: JSON.stringify({ actor: ACTOR }),
    },
  );
  return checkResponse(resp, `delete _rowid=${rowId}`);
}

function printRows(rows, title) {
  console.log(title);
  if (!rows.length) {
    console.log("  (empty)");
    return;
  }
  for (const r of rows) {
    console.log(
      `  _rowid=${r._rowid} · ${r.sku} · ${r.name} · price=${r.price} · stock=${r.stock}`,
    );
  }
}

console.log(`Target: ${DATABASE}/${TABLE}\n`);

const before = await listRows();
printRows(before.rows, `Current rows (${before.total} total):`);

const newRow = await insertRow({
  sku: "SKU-004",
  name: "Webcam HD",
  price: 199.0,
  stock: 30,
});
console.log(`\nInserted _rowid=${newRow._rowid} · ${newRow.name}`);

const middle = await listRows();
printRows(middle.rows, `\nAfter insert (${middle.total} total):`);

const targetId = before.rows[0]?._rowid ?? newRow._rowid;
await deleteRow(targetId);
console.log(`\nDeleted _rowid=${targetId}`);

const after = await listRows();
printRows(after.rows, `\nAfter delete (${after.total} total):`);
console.log("\nDone.");
console.log("Next: node tutorial/js/03_delete_table.mjs");

File: tutorial/02_read_insert_and_delete.py
"""
示例 2：读取表、插入新行、删除指定行

前置条件：
  - 已运行 01_create_table_and_write.py（表 tutorial_products 存在）
  - 本地服务已启动

运行：
  python tutorial/02_read_insert_and_delete.py
"""
from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

import requests

from config import ACTOR, DATABASE, api_url, check_response, site_headers, write_headers

TABLE = "tutorial_products"


def list_rows(*, limit: int = 20) -> dict:
    resp = requests.get(
        api_url(f"/api/databases/{DATABASE}/tables/{TABLE}/rows"),
        headers=site_headers(),
        params={"limit": limit, "order": "asc"},
        timeout=30,
    )
    return check_response(resp, "读取表")


def insert_row(row: dict) -> dict:
    resp = requests.post(
        api_url(f"/api/databases/{DATABASE}/tables/{TABLE}/rows"),
        headers=write_headers(),
        json={"actor": ACTOR, **row},
        timeout=30,
    )
    return check_response(resp, "插入新行")


def delete_row(row_id: int) -> dict:
    resp = requests.delete(
        api_url(f"/api/databases/{DATABASE}/tables/{TABLE}/rows/{row_id}"),
        headers=write_headers(),
        json={"actor": ACTOR},
        timeout=30,
    )
    return check_response(resp, f"删除 _rowid={row_id}")


def print_rows(rows: list[dict], title: str) -> None:
    print(title)
    if not rows:
        print("  (空表)")
        return
    for r in rows:
        print(
            f"  _rowid={r.get('_rowid')} · {r.get('sku')} · "
            f"{r.get('name')} · price={r.get('price')} · stock={r.get('stock')}"
        )


def main() -> None:
    print(f"目标: {DATABASE}/{TABLE}\n")

    # 1. 读取（无需 Write Key）
    before = list_rows()
    print_rows(before["rows"], f"当前数据（共 {before['total']} 行）:")

    # 2. 插入一行（需 Write Key）
    new_row = insert_row(
        {"sku": "SKU-004", "name": "Webcam HD", "price": 199.0, "stock": 30}
    )
    print(f"\n已插入新行 _rowid={new_row.get('_rowid')} · {new_row.get('name')}")

    # 3. 再次读取
    middle = list_rows()
    print_rows(middle["rows"], f"\n插入后（共 {middle['total']} 行）:")

    # 4. 删除第一行（按 _rowid）
    target_id = before["rows"][0]["_rowid"] if before["rows"] else new_row["_rowid"]
    delete_row(int(target_id))
    print(f"\n已删除 _rowid={target_id}")

    # 5. 最终状态
    after = list_rows()
    print_rows(after["rows"], f"\n删除后（共 {after['total']} 行）:")
    print("\n完成。")
    print("下一步: python tutorial/03_delete_table.py")


if __name__ == "__main__":
    main()
